src/libraries/System.Text.Json/tests/Serialization/CustomConverterTests/CustomConverterTests.Object.cs

```diff
@@ -311,6 +311,72 @@ public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOp
             }
         }
 
+        private class PrimitiveConverter : JsonConverter<object>
+        {
+            public override bool CanConvert(Type typeToConvert)
+                => true;
+
+            public override object Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options)
+            {
+                if (reader.TokenType == JsonTokenType.True)
+                {
+                    return true;
+                }
+
+                if (reader.TokenType == JsonTokenType.False)
+                {
+                    return false;
+                }
+
+                if (reader.TokenType == JsonTokenType.Number)
+                {
+                    if (reader.TryGetInt32(out int i))
+                    {
+                        return i;
+                    }
+
+                    return reader.GetDouble();
+                }
+
+                if (reader.TokenType == JsonTokenType.String)
+                {
+                    if (reader.TryGetDateTime(out DateTime datetime))
+                    {
+                        return datetime;
+                    }
+
+                    return reader.GetString();
+                }
+
+                throw new JsonException();
+            }
+
+            public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOptions options)
+            {
+                throw new InvalidOperationException("Should not get here.");
+            }
+        }
+        private class ClassWithPrimitives
+        {
+            [JsonConverter(typeof(PrimitiveConverter))]
+            public int MyInt { get; set; }
+            [JsonConverter(typeof(PrimitiveConverter))]
+            public bool MyBool { get; set; }
+            [JsonConverter(typeof(PrimitiveConverter))]
+            public string MyString { get; set; }
+        }
+
+        [Fact]
+        public static void ClassWithPrimitivesObjectConverterDeserialize()
+        {
+            const string json = @"{""MyInt"":123,""MyBool"":true,""MyString"":""Hello""}";
+
+            var obj = JsonSerializer.Deserialize<ClassWithPrimitives>(json);
+            Assert.Equal(123, obj.MyInt);
+            Assert.True(obj.MyBool);
+            Assert.Equal("Hello", obj.MyString);
+        }
```

Comment:
Please test serialization as well and ensure that the converter's write method is called.

This may require this line to be changed:

https://github.com/dotnet/runtime/blob/dc6ad2a89cde71714b87e42a159ed6264fa7093d/src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs#L334

```cs
if (type != TypeToConvert && IsInternalconverter)
```

Reply:
@layomia  This basically works for the tests in this PR but apparently at least `ExtensionDataTest.CustomObjectConverterInExtensionProperty` depends on the current behavior
``` cs
            public override void Write(Utf8JsonWriter writer, object value, JsonSerializerOptions options)
            {
                // Since we are converter for object, the string converter will be called instead of this.
                throw new InvalidOperationException();
            }
```
I can fix that `Write` implementation to work as expected. But don't know if this change is breaking anyone else.

Reply:
I think the premise that any `JsonConverter<object>` is an internal converter and shouldn't have it's `Write` method called is a wrong one. I don't imagine a lot of people depend on the behavior of the `Write` method not being called. I'd fix the test - let's see what the change looks like.